src/workerd/io/worker.c++

```diff
@@ -3444,49 +3439,67 @@ Worker::Actor::Actor(const Worker& worker,
 }
 
 void Worker::Actor::ensureConstructed(IoContext& context) {
-  KJ_IF_SOME(info, impl->classInstance.tryGet<Worker::Impl::ActorClassInfo*>()) {
-    context.addWaitUntil(context
-                             .run([this, &info = *info](Worker::Lock& lock) {
-      jsg::Lock& js = lock;
-
-      kj::Maybe<jsg::Ref<api::DurableObjectStorage>> storage;
-      KJ_IF_SOME(c, impl->actorCache) {
-        storage = impl->makeStorage(lock, worker->getIsolate().getApi(), *c);
-      }
-      auto handler = info.cls(lock,
-          jsg::alloc<api::DurableObjectState>(cloneId(),
-              jsg::JsRef<jsg::JsValue>(
-                  js, KJ_ASSERT_NONNULL(lock.getWorker().impl->ctxExports).addRef(js)),
-              kj::mv(storage), kj::mv(impl->container)),
-          KJ_ASSERT_NONNULL(lock.getWorker().impl->env).addRef(js));
-
-      // HACK: We set handler.env to undefined because we already passed the real env into the
-      //   constructor, and we want the handler methods to act like they take just one parameter.
-      //   We do the same for handler.ctx, as ExecutionContext related tasks are performed
-      //   on the actor's state field instead.
-      handler.env = js.v8Ref(js.v8Undefined());
-      handler.ctx = kj::none;
-      handler.missingSuperclass = info.missingSuperclass;
-
-      impl->classInstance = kj::mv(handler);
-    }).catch_([this](kj::Exception&& e) {
-      auto msg = e.getDescription();
-
-      if (!msg.startsWith("broken."_kj) && !msg.startsWith("remote.broken."_kj)) {
-        // If we already set up a brokenness reason, we shouldn't override it.
-
-        auto description = jsg::annotateBroken(msg, "broken.constructorFailed");
-        e.setDescription(kj::mv(description));
-      }
-
-      impl->constructorFailedPaf.fulfiller->reject(kj::cp(e));
-      impl->classInstance = kj::mv(e);
-    }));
-
+  KJ_IF_SOME(info, impl->classInstance.tryGet<ActorClassInfo*>()) {
+    context.addWaitUntil(ensureConstructedImpl(context, *info));
     impl->classInstance = Impl::Initializing();
   }
 }
 
+kj::Promise<void> Worker::Actor::ensureConstructedImpl(IoContext& context, ActorClassInfo& info) {
+  InputGate::Lock inputLock = co_await impl->inputGate.wait();
+
+  bool containerRunning = false;
+  KJ_IF_SOME(c, impl->container) {
+    // We need to do an RPC to check if the container is running.
+    // TODO(perf): It would be nice if we could have started this RPC earlier, e.g. in parallel
+    //   with starting the script, and also if we could save the status across hibernations. But
+    //   that would require some refactoring, and this RPC should (eventally) be local, so it's
+    //   not a huge deal.
+    auto status = co_await c.statusRequest(capnp::MessageSize{4, 0}).send();
```

Comment:
Can we get away with delaying the await to the first read rather than blocking while this request is completing?

Reply:
No because it's a synchronous property, so it has to already be there when the app reads it.

Almost all apps will read it immediately anyway.